### DISCUSSION TITLE
Has anyone successfully connected the keypad touch?

### DISCUSSION THREAD

**Discussion by nextdoor-coder** [hsakoh/switchbot-mqtt#22] [Q&A] [answered]
I was able to connect the control panel, but the sensors are not updating. 
Thanks in advance for any advice
![keypad](https://github.com/hsakoh/switchbot-mqtt/assets/67520021/e7072994-5559-4a84-b974-dbe6b884bc1b)

**Comment by hsakoh** [comment]
Are you able to enumerate the devices but having trouble retrieving the device status?

According to the API specifications, "Keypad Touch (MUST enable Cloud Service first)" is required.
https://github.com/OpenWonderLabs/SwitchBotAPI?tab=readme-ov-file#get-device-status
Have you enabled the Cloud Service?

Try setting the add-on's log level to trace, restart the add-on, and then check the responses during startup and polling. If you can provide the logs, it might give us more information to work with.

**Reply by hsakoh** [reply]
I've rechecked the specifications of the Keypad Touch API/Webhook, and it seems that the sensor values (fields in terms of the add-on) eventName, commandId, result can only be received via Webhook. Have you enabled the Webhook?

DeviceStatus
https://github.com/OpenWonderLabs/SwitchBotAPI?tab=readme-ov-file#keypad-touch-1
Webhook
https://github.com/OpenWonderLabs/SwitchBotAPI?tab=readme-ov-file#keypad-touch-3

Keypad Touch mainly relies on commands, and the fields do not provide particularly useful values. Ignoring the lack of updates or deleting the corresponding fields from the Ingress page could be good options.

Although it doesn't usually get updated and doesn't mean much, DeviceId and DeviceType were in DeviceStatus. 
These two fields were not polled due to errors in the master data.
I've updated the master data and released v1.0.16. 
Please remove the MQTT device, delete and re-acquire the device on the Ingress page, and restart the add-on.

**Reply by nextdoor-coder** [reply]
Thanks! I have updated to release v1.0.17 and DeviceId and DeviceType dispalyed now correctly.

I have also enabled Webhook, in the log i can see the following:
17:49:57 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Send GET,devices/xxxxxxxxxxx/status
17:49:58 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Response POST,https://api.switch-bot.com/v1.1/webhook/setupWebhook,{"statusCode":100,"body":{},"message":"success"}
17:49:58 info: SwitchBotMqttApp.Services.WebhookService[0] start listen http://192.168.88.11:8098/webhook/webhook
17:49:58 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Response GET,https://api.switch-bot.com/v1.1/devices/xxxxxxxxxxx/status,{"statusCode":100,"body":{"deviceId":"xxxxxxxxxxx","deviceType":"Hub 2","hubDeviceId":"xxxxxxxxxxx","humidity":61,"temperature":24.2,"lightLevel":2,"version":"V2.0-1.2"},"message":"success"}
17:49:58 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Send GET,devices/zzzzzzzzzz/status
17:49:59 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Response GET,https://api.switch-bot.com/v1.1/devices/zzzzzzzzzz/status,{"statusCode":100,"body":{"deviceId":"zzzzzzzzzz","deviceType":"Keypad Touch","hubDeviceId":"xxxxxxxxxxx"},"message":"success"}
17:49:59 warn: SwitchBotMqttApp.Services.MqttCoreService[0] unknown status paylod KeypadTouch,hubDeviceId,"xxxxxxxxxxx"
17:49:59 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Send GET,scenes
17:50:00 trce: SwitchBotMqttApp.Logics.SwitchBotApiClient[0] Response GET,https://api.switch-bot.com/v1.1/scenes,{"statusCode":100,"body":[],"message":"success"}

**Reply by hsakoh** [reply]
It's great that DeviceId and DeviceType can now be received.
(hubDeviceId is listed as an unknown payload item, but it's not very meaningful, so it will be added to the master data in the next release.)

Regarding webhook configuration, specifying a local IP is generally incorrect.
You need to set a URL that can be called from the SwitchBot server, which may require port forwarding.
(Alternatively, consider using Ngrok... Personally, I recommend using Ngrok.)

Specifically, please refer to the following issue and comments:
https://github.com/hsakoh/switchbot-mqtt/issues/17#issuecomment-2118721153

**Reply by nextdoor-coder** [reply]
Thanks for the hint. I did a quick test by using the Public IP and port forwarding. 
Webhook seems to work in general 

On Lock Pro and Hub 2 the "Webhook Lastupdate" and "timeOfSample" sensor gets updated, but not for Keypad Touch even with Cloud Service enabled

**Reply by balu79** [reply]
> Thanks for the hint. I did a quick test by using the Public IP and port forwarding. Webhook seems to work in general
> 
> On Lock Pro and Hub 2 the "Webhook Lastupdate" and "timeOfSample" sensor gets updated, but not for Keypad Touch even with Cloud Service enabled

Same for me unfortunately. DeviceId and DeviceType are displayed but all the sensors like Webhook Lastupdate and commandId are shown as unknown. I'm using Ngrok, Cloud Services are enabled and webhook is enabled.

**Comment by hsakoh** [comment, marked as answer]
@balu79 

I was able to test the Keypad Touch again. The API responses and Webhook reception results met the published specifications. The add-on also behaved as expected.

### Explanation of the Fields Related to the Keypad Touch

1. Fields updated by the GetDeviceStatus API response:
These fields are updated when values are obtained through manual polling, automatic polling, or when the add-on is started (at the beginning of MqttCoreService). The three fields are `deviceId`, `deviceType`, and `Status Lastupdate`.

2. Fields updated by the Webhook:
The six fields are `deviceId`, `deviceType`, `eventName`, `commandId`, `result`, and `timeOfSample`.

3. Conditions for Webhook Trigger:
A Webhook is triggered when a key is created or deleted.

### Bug in Add-on Version v1.0.15 and Earlier
In versions prior to v1.0.15, due to an error in the master data, `deviceId` and `deviceType` were not being updated through polling.

### Points to Confirm
1. Is the cloud service for the Keypad Touch enabled?
2. Is the add-on's Webhook correctly configured?
   It would be ideal if you could confirm the reception of Webhooks with other devices.

   You can also verify the configuration and reception from the add-on logs.
   ```
   hh:mm:dd info: SwitchBotMqttApp.Services.WebhookService[0] start listen https://xxxxxx.ngrok-free.app/webhook
   hh:mm:dd info: SwitchBotMqttApp.Controllers.WebhookController[0] {"eventType":"changeReport","eventVersion":"1",....}
   ```
   The first line is a log that occurs when the Webhook is active, indicating the URL registered with SwitchBot (this is an example using Ngrok).
   If you access this URL with a browser, you should receive a 405 Method Not Allowed response.
   (If this URL is not accessible from the internet, there may be an issue with the configuration.)

   The second line indicates that some Webhook has been received.

3. Is the Webhook checkbox for the Keypad Touch turned on in the device configuration page?
4. Are the checkboxes for `eventName`, `commandId`, `result`, and `timeOfSample` for the Keypad Touch turned on in the device configuration page?
5. Are the conditions for triggering the Webhook for the Keypad Touch met?
   A Webhook is triggered only when a key is created or deleted.
   (Webhook generation has been confirmed when using the API for these operations. It is not yet confirmed whether it triggers when performed via the app.)
6. Is the add-on version up to date?
7. Please try the following steps to clean up the devices on Home Assistant:
    1. Delete the detected Keypad Touch MQTT device from the devices page in Home Assistant.
    2. Delete the Keypad Touch from the device configuration page and save the changes.
    3. On the device configuration page, click the "Fetch Devices By SwitchBotApi" button to retrieve the Keypad Touch device again.
    4. On the device configuration page, turn on the checkboxes for Enable, Polling, Webhook, and the relevant fields for the Keypad Touch, and save the changes.
    5. Restart the add-on, or restart (stop and start) the MqttCoreService from the service management page.

**Reply by balu79** [reply]
Thanks so much for all the time you took to lay this out in so much detail, really fool proof, much appreciated!

And now I can also confirm it works fine for me as well. I didn't know the trigger of the webhook for the keypad is "only" triggered when a key is created or deleted. I now tested that and could see the sensors all updating fine based on the webhook.
